dialogue-core/src/main/java/com/palantir/dialogue/core/QueuedChannel.java

```diff
@@ -0,0 +1,123 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.core;
+
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.common.util.concurrent.SettableFuture;
+import com.palantir.conjure.java.api.errors.QosException;
+import com.palantir.dialogue.Channel;
+import com.palantir.dialogue.Endpoint;
+import com.palantir.dialogue.Request;
+import com.palantir.dialogue.Response;
+import java.util.Deque;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.Executor;
+import org.immutables.value.Value;
+
+final class QueuedChannel implements Channel {
+
+    private static final Executor DIRECT = MoreExecutors.directExecutor();
+    private final Deque<DeferredCall> queuedCalls = new ConcurrentLinkedDeque<>();
+    private final LimitedChannel delegate;
+
+    QueuedChannel(LimitedChannel delegate) {
+        this.delegate = delegate;
+    }
+
+    /**
+     * Enqueue the call and try to schedule as many queued tasks as possible.
+     */
+    @Override
+    public ListenableFuture<Response> execute(Endpoint endpoint, Request request) {
+        DeferredCall components = ImmutableDeferredCall.of(endpoint, request, SettableFuture.create());
+
+        if (!queuedCalls.offer(components)) {
+            throw QosException.unavailable();
+        }
+
+        schedule();
+
+        return components.response();
+    }
+
+    /**
+     * Try to schedule as many tasks as possible. Called when requests are submitted and when they complete.
+     */
+    private void schedule() {
+        while (scheduleNextTask()) {
+            // Do nothing
+        }
+    }
+
+    /**
+     * Get the next call and attempt to execute it. If it is runnable, wire up the underlying future to the one
+     * previously returned to the caller. If it is not runnable, add it back into the queue. Returns true if more
+     * tasks may be able to be scheduled, and false otherwise.
+     */
+    private boolean scheduleNextTask() {
+        DeferredCall components = queuedCalls.poll();
+        if (components == null) {
+            return false;
+        }
+
+        Optional<ListenableFuture<Response>> response =
+                delegate.maybeExecute(components.endpoint(), components.request());
+
+        if (response.isPresent()) {
+            Futures.addCallback(response.get(), new ForwardAndSchedule(components.response()), DIRECT);
+            return true;
+        } else {
+            queuedCalls.addFirst(components);
+            return false;
```

Comment:
this doesn't work as written. if my call doesn't execute the first time around, then it ends up in the queue and will only run the next time i submit a call on this channel, right?

Reply:
ah, you have the schedule thing below, i see. I think this whole contraption is a lot easier with a dedicated scheduling thread and a blocking queue as described above.

Reply:
issue mentioned above. the scheduling thread can't always do the next bit of work, so you either have to
1. schedule it to run periodically
2. spin and keep trying to execute the next request
3. signaling to the scheduling thread that it needs to run (eg: when a request completes)
4. have the underlying channel tell the scheduling thread to retryAfter some period of time

Reply:
can we talk over VTC, I don't think I understand the flow you have in mind here.

Reply:
Yeah, you'd end up with bursts of request sent every 100ms rather than a smooth request rate. I'm going to merge this for now and can revisit going forward.

Reply:
👍